### PR TITLE
patch nvrtc API for cuda TK >= 11.1

### PR DESCRIPTION
**Pull request by jjsjann123**
CUDA TK >= 11.1 provides ptxjitcompiler that emits SASS instead of PTX.
1. This gives better backward-compatibility that allows future TK to work with older driver, which might not necessarily be able to load generated PTX through JIT compile and would error out at runtime;
https://docs.nvidia.com/deploy/cuda-compatibility/#using-ptx
2. Meanwhile, SASS doesn't provide good future compatibility, so for unsupported arch, we fallback to PTX to support future device.
https://docs.nvidia.com/deploy/cuda-compatibility/index.html#cubin-compatibility

### CODE REVIEW

**Comment by jjsjann123** [comment]
cc'ing @ptrblck @xwang233

File: torch/csrc/jit/codegen/cuda/executor_utils.cpp [hunk 1]
```diff
@@ -280,8 +280,13 @@ NvrtcFunction nvrtcCompile(
 
 #ifdef __HIP_PLATFORM_HCC__
   std::vector<const char*> args = {"--std=c++14"};
+#else
+#if CUDA_VERSION == 11020
```
Comment (xwang233) [inline torch/csrc/jit/codegen/cuda/executor_utils.cpp:284]: Do we need to change those macros to ">="?

Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/executor_utils.cpp:284]: I was hoping this is a one time thing. Let me double check it with those compiler guys.

**Comment by ngimel** [comment]
So is it an option to use sass, or a bug that forces you to use sass? Comments in the PR imply the latter, description the former. Also, description says >=11.1, and PR uses ==11.2. How exactly using sass gives better backward compatibility, please describe in which situations it would matter.

**Comment by jjsjann123** [comment]
> As mentioned in the PTX section, the compilation of PTX to device code lives along with the CUDA driver, hence the generated PTX might be newer than what is supported by the driver on the deployment system.
This is from the doc I linked. So it is an error thrown by nvrtc (if you have older driver that doesn't support the PTX generated by newer library)

The new API is available in CUDA 11.1, we haven't run into issues in our container with 11.1 but only 11.2, hence the fix only targeted that. (I'm trying to confirm this with compiler guys)

**Comment by jjsjann123** [comment]
I was being overly optimistic about compatibility issues here. :stuck_out_tongue_closed_eyes: 
I'll update the PR with comment inline. Will ping you when it's ready again.

**Comment by jjsjann123** [comment]
Looks like CI is green now. Try my luck with @ngimel again.

**Review by xwang233 — APPROVED** [review]
LGTM

**Comment by jjsjann123** [comment]
ping @ngimel :pray:

File: torch/csrc/jit/tensorexpr/cuda_codegen.cpp [hunk 2]
```diff
@@ -99,6 +100,8 @@ static void getMajorMinor(
   }
   major = dev_version.first;
   minor = dev_version.second;
+
+  return (major != nvrtc_version.first) || (minor != nvrtc_version.second);
```
Comment (ngimel) [inline torch/csrc/jit/tensorexpr/cuda_codegen.cpp:104]: Are you comparing the right things here? `major` is like 8 for ampere, and nvrtc_version.first is 11?

File: torch/csrc/jit/codegen/fuser/cuda/fused_kernel.cpp [hunk 3]
```diff
@@ -64,6 +65,8 @@ void getMajorMinor(const cudaDeviceProp* const prop, int& major, int& minor) {
     major = 8;
     minor = 0;
   }
+
+  return (major != prop->major) || (minor != prop->minor);
```
Comment (ngimel) [inline torch/csrc/jit/codegen/fuser/cuda/fused_kernel.cpp:69]: So for 11.1 this is always `false`, major == prop->major, same for minor. This means here https://github.com/pytorch/pytorch/pull/50319/files#diff-5e44e2a0e9a686349e4152cd3f1255ae55c3c208b22112f194144d8c6612f61aR296 we are going to ptx - but it was supposed to be the other way around?

File: aten/src/ATen/cuda/nvrtc_stub/ATenNVRTC.h [hunk 4]
```diff
@@ -29,13 +29,16 @@ namespace at { namespace cuda {
...
 
+#if CUDA_VERSION >= 11010
 #define AT_FORALL_NVRTC(_)                       \
   _(nvrtcVersion)                                \
   _(nvrtcAddNameExpression)                      \
   _(nvrtcCreateProgram)                          \
   _(nvrtcDestroyProgram)                         \
   _(nvrtcGetPTXSize)                             \
   _(nvrtcGetPTX)                                 \
+  _(nvrtcGetCUBINSize)                           \
```
[the remark this answers is not in the document]
Comment (jjsjann123) [reply aten/src/ATen/cuda/nvrtc_stub/ATenNVRTC.h]: My poor macro can't figure out how to nest `#if` inside `#define`. Would this be ok:
https://github.com/pytorch/pytorch/pull/50319/files#diff-ea7572ef8857692509ae9f2810d9dd62a9b3157ef8c945b3dd68aa7ef5cc8dd9R57-R66

File: torch/csrc/jit/codegen/cuda/executor_utils.cpp [hunk 5]
```diff
@@ -262,7 +262,10 @@ NvrtcFunction nvrtcCompile(
   const auto prop = at::cuda::getCurrentDeviceProperties();
 
   int major = 0, minor = 0;
-  getMajorMinor(prop, major, minor);
+#if CUDA_VERSION >= 11010
+  const bool supported_arch =
+#endif
+      getMajorMinor(prop, major, minor);
```
[the remark this answers is not in the document]
Comment (jjsjann123) [reply torch/csrc/jit/codegen/cuda/executor_utils.cpp]: so how about `codegenOutputQuery`?

File: torch/csrc/jit/tensorexpr/cuda_codegen.cpp (hunk 2 above)
Comment (jjsjann123) [reply torch/csrc/jit/tensorexpr/cuda_codegen.cpp]: good catch.

File: torch/csrc/jit/codegen/fuser/cuda/fused_kernel.cpp (hunk 3 above)
Comment (jjsjann123) [reply torch/csrc/jit/codegen/fuser/cuda/fused_kernel.cpp]: Good catch.

**Review by ngimel — APPROVED** [review]
Ok, now that looks good, thanks

**Pull request closed without merging**